### ISSUE TITLE
Experimental Code - CTC Loss issue.

### ISSUE THREAD

**Issue by iiSeymour** [idiap/fast-transformers#5]
Firstly, great paper and accompanying repo, thanks.

I am evaluating the fast-transformer on a problem similar to section 4.3 in the paper where I have a non-autoregressive task trained using CTCLoss. I define my new model as per below but run into a problem during training where the loss converges to 1.0 and gets stuck?

```python
class Model(Module):

    def __init__(self, classes=5, stride=3, n=5, d=64):
        super().__init__()

        Transformer = TransformerEncoderBuilder.from_kwargs(
            n_layers=n,
            n_heads=n,
            query_dimensions=d,
            value_dimensions=d,
            feed_forward_dimensions=256,
            activation='gelu',
            dropout=0.05,
            attention_type = "linear",
        )

        self.layers = Sequential(
            # N, C, W
            Conv1d(1, n * d, self.k, padding=4, stride=stride),
            Transpose(1, 2),
            # N, W, C
            Transformer.get(),
            Linear(n * d, classes),
            LogSoftmax(2),
        )

    def forward(self, x):
        return self.layers(x)
```

Is the code used in section 4 also going to be made available?

Thanks

**Comment by apoorv2904** [comment]
Hi,

Thank you for your interest in our work. Yes, we will release a separate repo with our experiments.  Though for the above code snippet, I can see that the position embeddings are missing.  Try adding those to see if the loss converges.

Thanks,
Apoorv

**Comment by iiSeymour** [comment]
Ah yes of course, can you explain how you handled the positional embeddings for ASR?

**Comment by apoorv2904** [comment]
We use fixed sinusoidal positional embeddings that are concatenated to the input speech features.
The concatenated features are then linearly projected and given as input to the transformer.

Following is the PositionalEmbedding class that takes as input speech features of shape N x T x E and returns concatenated features. The output has the shape N x T x (E + d_model)

```
class PositionalEmbedding(nn.Module):

    def __init__(self, d_model, dropout=0.0, max_len=5000):
        super(PositionalEmbedding, self).__init__()
        self.dropout = nn.Dropout(p=dropout)

        self.d_model = d_model
        pe = torch.zeros(max_len, d_model)
        position = torch.arange(0, max_len, dtype=torch.float).unsqueeze(1)
        div_term = torch.exp(torch.arange(0, d_model, 2).float() * (-math.log(10000.0) / d_model))
        pe[:, 0::2] = torch.sin(position * div_term)
        pe[:, 1::2] = torch.cos(position * div_term)
        pe = pe.unsqueeze(0)
        self.register_buffer('pe', pe) 

    def forward(self, x): 
        pos_embedding =  self.pe[:, :x.size(1), :]
        pos_embedding = torch.repeat_interleave(pos_embedding, x.shape[0], dim=0)
        x =  torch.cat([x, pos_embedding], dim=2)
        return self.dropout(x)
```

**Comment by angeloskath** [comment]
@iiSeymour Chris I am closing the issue, feel free to reopen it or submit a new one if you are still experiencing problems.

Cheers,
Angelos
